### ISSUE TITLE
TSS: Environment Variables are ignored

### ISSUE THREAD

**Issue by ari75** [ansible-collections/community.general#3358]
### Summary

Following the documentation of the TSS lookup plugin available [here](https://docs.ansible.com/ansible/latest/collections/community/general/tss_lookup.html), it should be possible to use Environment Variables to define the key parameters of the plugin:
  - base_url (env:TSS_BASE_URL)
  - username (env:TSS_USERNAME)
  - password (env:TSS_PASSWORD)

However these environment variables don't seems to be loaded at all by the plugin: [code](https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/tss.py)

It's at least what i can see when I compare with another lookup plugin: https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/cyberarkpassword.py
Line 83: CLIPASSWORDSDK_CMD = os.getenv('AIM_CLIPASSWORDSDK_CMD', '/opt/CARKaim/sdk/clipasswordsdk')

Thanks for your support

### Issue Type

Bug Report

### Component Name

tss

### Ansible Version

Ansible211 through AnsibleTower

### Configuration

```
 - name: Debug environment variables on ansible tower host
  local_action: ansible.builtin.shell echo TSS Environment Variables $HOSTNAME @ $TSS_BASE_URL, $TSS_USERNAME, $TSS_PASSWORD
  environment:
    TSS_BASE_URL: "{{thycotic_base_url}}"
    TSS_USERNAME: "{{thycotic_username}}"
    TSS_PASSWORD: "{{thycotic_key}}"

- name: Debug environment variables on target hosts
  ansible.builtin.shell: echo TSS Environment Variables $HOSTNAME @ $TSS_BASE_URL, $TSS_USERNAME, $TSS_PASSWORD
  environment:
    TSS_BASE_URL: "{{thycotic_base_url}}"
    TSS_USERNAME: "{{thycotic_username}}"
    TSS_PASSWORD: "{{thycotic_key}}"

- name: Load Thycotic secrets and debug them
  ansible.builtin.debug:
    msg: "Thycotic plugin found secrets: {{secret1}} and {{secret2}}"
  vars:
    secret1: "{{ lookup('community.general.tss', 1) }}"
    secret2: "{{ lookup('community.general.tss', 2) }}"
  environment:
    TSS_BASE_URL: "{{thycotic_base_url}}"
    TSS_USERNAME: "{{thycotic_username}}"
    TSS_PASSWORD: "{{thycotic_key}}"
  delegate_to: 127.0.0.1
```

### OS / Environment

RHEL8

### Steps to Reproduce

See description

### Expected Results

I expect the environment variables to work

### Actual Results

```
TASK [thycotic : Debug environment variables on ansible tower host] ************
changed: [myServer -> localhost] => {"changed": true, "cmd": "echo TSS Environment Variables $HOSTNAME @ $TSS_BASE_URL, $TSS_USERNAME, $TSS_PASSWORD", "delta": "0:00:00.005840", "end": "2021-09-10 16:36:05.325238", "rc": 0, "start": "2021-09-10 16:36:05.319398", "stderr": "", "stderr_lines": [], "stdout": "TSS Environment Variables myServer @ https://myTssServer/SecretServer/, myUsername, myPassword", "stdout_lines": ["TSS Environment Variables myServer @ https://myTssServer/SecretServer/, myUsername, myPassword"]}
TASK [thycotic : Debug environment variables on target hosts] ******************
changed: [localhost] => {"changed": true, "cmd": "echo TSS Environment Variables $HOSTNAME @ $TSS_BASE_URL, $TSS_USERNAME, $TSS_PASSWORD", "delta": "0:00:00.003763", "end": "2021-09-10 18:36:05.639825", "rc": 0, "start": "2021-09-10 18:36:05.636062", "stderr": "", "stderr_lines": [], "stdout": "TSS Environment Variables localhost @ https://myTssServer/SecretServer/, myUsername, myPassword", "stdout_lines": ["TSS Environment Variables localhost @ https://myTssServer/SecretServer/, myUsername, myPassword"]}
TASK [thycotic : Load Thycotic secrets and debug them] *************************
fatal: [myServer -> 127.0.0.1]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ lookup('community.general.tss', 1) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while running the lookup plugin 'community.general.tss'. Error was a <class 'ansible.errors.AnsibleError'>, original message: No setting was provided for required configuration plugin_type: lookup plugin: ansible_collections.community.general.plugins.lookup.tss setting: base_url "}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibullbot** [comment]
Files identified in the description:
* [`plugins/lookup/tss.py`](https://github.com/ansible-collections/community.general/blob/master/plugins/lookup/tss.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibullbot** [comment]
cc @amigus
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Ajpantuso** [comment]
What's being done in the `cyberarkpassword` plugin is unnecessary as the options are read directly from the `DOCUMENTATION` string (including the `ENV` names). What may be your issue is the use of the `environment` keyword (See [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_environment.html#setting-the-remote-environment). Those variables when defined at the task or playbook level do not influence `Ansible` or `plugins`.

**Comment by felixfontein** [comment]
>     * base_url (env:TSS_BASE_URL)
> 
>     * username (env:TSS_USERNAME)
> 
>     * password (env:TSS_PASSWORD)
> 
> 
> However these environment variables don't seems to be loaded at all by the plugin: [code](https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/tss.py)

They don't need to be handled by the code (and in fact should not be), since Ansible's config manager takes care of them: https://github.com/ansible/ansible/blob/devel/lib/ansible/config/manager.py#L496

> It's at least what i can see when I compare with another lookup plugin: https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/cyberarkpassword.py
> Line 83: CLIPASSWORDSDK_CMD = os.getenv('AIM_CLIPASSWORDSDK_CMD', '/opt/CARKaim/sdk/clipasswordsdk')

That plugin isn't yet using the config manager, so it has to do everything manually.

> ```
> - name: Load Thycotic secrets and debug them
>   ansible.builtin.debug:
>     msg: "Thycotic plugin found secrets: {{secret1}} and {{secret2}}"
>   vars:
>     secret1: "{{ lookup('community.general.tss', 1) }}"
>     secret2: "{{ lookup('community.general.tss', 2) }}"
>   environment:
>     TSS_BASE_URL: "{{thycotic_base_url}}"
>     TSS_USERNAME: "{{thycotic_username}}"
>     TSS_PASSWORD: "{{thycotic_key}}"
>   delegate_to: 127.0.0.1
> ```

You have to define the environment variables when running the playbook. Here you are defining them for the `ansible.builtin.debug` module (which is in fact an action plugin and won't use them at all), and will only be used during module execution - but not during preparation of the task.

**Comment by Ajpantuso** [comment]
@ari75 were you able to successfully use environment variables for the tss lookup plugin based on the recommendations here?

**Comment by ari75** [comment]
Hello,

Thank you very much for the explanation! It makes sense yes.

I didn't have time to test it yet, I will let you inform as soon as I'll have more feedback to provide.

Please let the issue pending meanwhile.

Cheers

**Comment by ansibullbot** [comment]
cc @endlesstrax
[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by ari75** [comment]
Hello,

Finally I could do some tests. As we are using Ansible Tower, we went for the option to use a custom Ansible Credential, as specified here: https://docs.ansible.com/ansible-tower/latest/html/userguide/credential_types.html. This one will in turn set up the 3 parameters as Environment Variable.

I can now confirm that these parameters are at least processed by the plugin:
`changed: [myserver -> localhost] => 
{
	"changed": true,
	"cmd": "echo TSS Environment Variables $HOSTNAME @ $TSS_BASE_URL, $TSS_USERNAME, $TSS_PASSWORD",
	"delta": "0:00:00.004395",
	"end": "2021-10-01 12:33:29.599253",
	"rc": 0,
	"start": "2021-10-01 12:33:29.594858",
	"stderr": "",
	"stderr_lines": [],
	"stdout": "TSS Environment Variables myserver @ https://tss.server.domain/, myusername, mypassword",
	"stdout_lines": ["TSS Environment Variables myserver @ https://tss.server.domain/, myusername, mypassword"]
}`

However the plugin fails with this error:
`fatal: [myserver -> 127.0.0.1]: FAILED! => 
{
	"msg": "An unhandled exception occurred while templating '{{ lookup('community.general.tss', mysecretid) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while running the lookup plugin 'community.general.tss'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Secret Server lookup failure: Login failed."
}`

To clarify, I made the assumption that the TSS SDK parameters were expected, as the TSS SDK is mentionned as a dependency for this plugin to work. Therefore I tryied using as username the "sdk_client_rule" and as password "sdk_client_key" that I used for testing the SDK CLI. These parameters work for the CLI, but not for the Plugin. I also tried using the thycotic "username" bound to this rule without more success.

Could you please explain which values actually configured in Thycotics are expected here?
Is there any way to add debugging logs for the plugin or for the underlying SDK?

Thanks for your support

**Comment by EndlessTrax** [comment]
If you are using Ansible Tower, there is a [TSS Credential Plugin](https://github.com/ansible/awx/blob/devel/awx/main/credential_plugins/tss.py). Not sure if you are aware of that. 

As for your login issue, is `sdk_client_rule` an application user, local, or AD user?

**Comment by ari75** [comment]
hi @EndlessTrax ,

### Regarding the TSS Credential Plugin for Ansible Tower
I was not aware of this TSS Credential plugin for Ansible Tower, thanks for the hint.

Do I understand correctly that:
- One must build the custom plugin himself and install it in each Ansible Tower venv where it shall be used?
- One credential must created for each secret to be retrieved from Thycotic SS? 

In this case, this would be quite unpractical to use for us from an operation perspective.

Is it possible to provide instead a list of key pair with secretId and variable name so that all variables could be loaded at once?

We were so far planning to store all the secretId in the inventory per environment and simply retrieving the values at the start of the ansible playbook using this ansible lookup plugin, but if we had a way to configure the list of variables and secretIds and load them all at once before starting the ansible playbook, it would be ideal.

**Regarding the Ansible Thycotic Lookup-Plugin**
About your question, we got a sdk_client_rule with a preestablished key generated for the connection with Ansible Tower. This sdk_client_rule is related to an application user. We initialized the connection already in the SDK and we can confirm that we can obtain the secrets through the CLI. We expect the Ansible TSS Plugin to somehow reuse this connection to obtain the secrets but we may be wrong.

Do we understand correctly that the authentication of this Plugin to the TSS can work either:
- Using a personal username and password, which may not be ideal or even acceptable in term of security.
- Using a Token. The only way we are aware of to obtain such a token is through a direct call in the CLI. It would also here not 
be very convenient if we would need to manually do this and update the Ansible Tower custom credential each time we need to run the playbook. Is there any other way to obtain such token? Or is it another type of Token meaned here?

The authentication process use-cases are not described in the plugin documentation and we don't know what is possible and what would be the best practice.

Thanks for your support.

**Comment by zcthompson** [comment]
It appears the current implementation of the `PasswordGrantAuthorizer` and `DomainPasswordGrantAuthorizer` in `plugins/lookup/tss.py` does not correctly fall back to environment variables (`TSS_USERNAME`, `TSS_PASSWORD`, etc.) when `username` and `password` are not explicitly passed.

This leads to unexpected authentication errors when using environment-based credential injection (common in CI/CD or vault setups).

### Proposed Fix
Introduce a utility method `_with_env_fallback()` to check for the key in `server_parameters` and fall back to the respective environment variable. Then refactor `_get_authorizer()` in `TSSClientV1` to consistently use this.

### Suggested Implementation
Add near the top of the file:
```python
def _with_env_fallback(server_parameters, key, env_var):
    return server_parameters.get(key) or os.getenv(env_var)
```
Then update _get_authorizer():
```
@staticmethod
def _get_authorizer(**server_parameters):
    base_url = _with_env_fallback(server_parameters, "base_url", "TSS_BASE_URL")
    username = _with_env_fallback(server_parameters, "username", "TSS_USERNAME")
    password = _with_env_fallback(server_parameters, "password", "TSS_PASSWORD")
    domain = _with_env_fallback(server_parameters, "domain", "TSS_DOMAIN")
    token = _with_env_fallback(server_parameters, "token", "TSS_TOKEN")
    token_path_uri = _with_env_fallback(server_parameters, "token_path_uri", "TSS_TOKEN_PATH_URI")

    if token:
        return AccessTokenAuthorizer(token)
    if domain:
        return DomainPasswordGrantAuthorizer(base_url, username, domain, password, token_path_uri)
    return PasswordGrantAuthorizer(base_url, username, password, token_path_uri)
```

**Comment by bcoca** [comment]
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/parameters.py#L833 not usable in this case?

**Comment by zcthompson** [comment]
@bcoca, This was just suggested to correct the code in a manner where it will fall back and make it a more simplistic user experience rather than hunting down things like I have over this for a use case where we really had to create a python script and use the python-tss-sdk (so this is mimicked) instead of using the lookup for tss all together as it did not really to appear to be respecting the authorizer portion after the username/password/base_url were passed as extra vars for simple username/password auth and kept going back to the token path for oauth.  At any rate, it was just a suggested fix to the plugin, I haven't taken the time to submit it in a PR fashion but it seems like it could remove some headache for those less familiar with some of these caveats but yes, that could be used in your example as well I think to achieve it in the env.

**Comment by bcoca** [comment]
But it departs from the standard precedence rules in ansible, that will be confusing to many users. I also consider updating the documentation (which is also the source for option handling) a much simpler change than adding non standard option handling code.

**Comment by zcthompson** [comment]
Possibly, the approach thought here was to make it function as intended as it's documented, I see your point of view, the code was a suggestion and open to criticism, but it also alleviates another issue we encountered which was SSL errors that weren't actually SSL errors because it was not respecting items in the authorizer as it should for username/password, even with environment vars passed or username/pw pulled as extra_vars.  Thanks for the feedback on this

**Comment by russoz** [comment]
Hi @ari75 @zcthompson 

From what I read, a summary of the proposal is: changing the usual precedence order between environment variables and ansible.cfg variables. 

If I got that right, it sounds like a bad idea to have one specific plugin work with a different set of rules of all the other plugins. 

Putting that together with the fact that the original claim (that variables were being ignored) has been disproved, I am leaning towards closing this issue, unless there is something more to it that I missed. 

needs_info
